design-proposals/storage-class-migration/storage-class-migration.md

```diff
@@ -0,0 +1,267 @@
+
+According to the 'ReclaimPolicySourcePvc' in both scenarios, KubeVirt will determine how to handle the source PVCs. The source PVCs will be automatically removed if the migration is successful, and the only option available right now is "DeleteReclaimPolicySourcePvc." The design debate is the time to include additional potential policies.
+
+However, the core idea is that it is KubeVirt responsability to handle fate of the source PVCs as it is the only component able to orchestrate and set the migration as successfull.
+
+## Design details and motivation
+
+### Motivation on the approaches
+
+From libvirt side there a multiple ways to achieve the storage migration. It is possible to simply copy the storage on the new destination and then swap the original disk wit the new one. This approach is the most perfomant as the overhead is caused only by the storage copy.
+
+A different approach is to simultaneously move the storage and live-migrate the guest. When the destination storage is not available on the host where the VM is executing, this solution is required. Costs are greater since there is a storage migration overhead as well as a live VM migration overhead.
+
+As KubeVirt is designed today and due to Kubernetes design principles of volume pod immutability, it isn't possible to hotplug and unplug new volumes while the pod is running. KubeVirt already supports volume hotplug by attaching new volume to a dummy pod and then by bound mounting the storage in `virt-launcher` pod. Sadly, this method fails to address the issue of removing the old PVCs that were expressly mentioned in the virt-launcher pod.
+Declaring all volumes hotpluggable could be a potential remedy, but we prefer to avoid it because it has been shown to be fragile for some storage providers and opaque to Kubernetes.
+
+The only workable and generic solution for this restriction appears to be live VM migration with storage. Live migration involves the creation of a new virt-launcher pod by KubeVirt that has the destination PVCs attached. The source virt-launcher pod and source PVCs are removed after the migration is finished because they are no longer needed.
+
+This migrating to a new storage provider is an exceptional operations as very resource consuming. If in the future, the k8s community will consider the volume mutability in the pod spec, then the lighter approach could be taken into consideration.
+
+### Design
+
+This feature adds an additional controller for the CRD `Storage Migration` to virt-controller. Once the StorageMigration object is created then KubeVirt will update the VMI status with the source and destination volume pair.
+If the user specifies the storage migration classes, then the destination PVCs will be created before updating the VMI status.
```

Comment:
What name will be used for the PVCs created by KubeVirt?

Reply:
This is something I'd love to decide in the design discussion. But you are right I haven't specified it. I think this should be configurable by the user in the StorageMigration object. We could use a prefix and then append to it a random string or a list of possible names.

If you have other suggestion, please, don't hesitate.

Reply:
I added the `PrefixDestinationVolume` as I think is the easiest thing to implement as a first behavior

Reply:
I think it is desirable as migrating the storage is costly and the user probably wants to do only once. However, we need to check all the possible weird corner cases and conflicts. For example, what if I specify a source PVC in `MigrationVolumes` but at the same time its storage class is also in `MigrationStorageClass`. I think in this case, we need to thrown an error. A volume can only be either in `MigrationVolumes` or in `MigrationStorageClass`

Reply:
> what if I specify a source PVC in MigrationVolumes but at the same time its storage class is also in MigrationStorageClass.

Yes. That is the use case I tried to describe (probably too badly). 

> I think in this case, we need to thrown an error. A volume can only be either in MigrationVolumes or in MigrationStorageClass

I was thinking more in the direction of allowing such a 'conflict' and just treat it as a way of specifying the name of the destination volume explicitly (so it does not get automatically generated). E.g. assuming I have 3 volumes of the same storage class `v1, v2 and v3` and I care that `v1` gets renamed as `v4`. Then I'd specify the `MigrationStorageClass` and only one `MigrationVolume` with `src` as `v1` and `dest` as `v4`. The names for volumes `v2` and `v3` would be automatically generated using some random strings.

Reply:
That's a neat idea! I would allow also to specify a different reclaim policy.

Reply:
I integrated this into the proposal, please have a second look when you have time